server/middlewares/validators/videos/videos.ts

```diff
@@ -106,26 +108,130 @@ const videosAddValidator = getCommonVideoEditAttributes().concat([
+
+      return clearUploadFile(file.path)
+    }
+
+    res.locals.videoFileResumable = file
+
+    return next()
+  }
+]
+
+/**
+ * File is created in POST initialisation, and its body is saved as a 'metadata' field is saved by uploadx for later use.
+ * see https://github.com/kukhariev/node-uploadx/blob/dc9fb4a8ac5a6f481902588e93062f591ec6ef03/packages/core/src/handlers/uploadx.ts#L41
+ *
+ * Uploadx doesn't use next() until the upload completes, so this middleware has to be placed before uploadx
+ * see https://github.com/kukhariev/node-uploadx/blob/dc9fb4a8ac5a6f481902588e93062f591ec6ef03/packages/core/src/handlers/base-handler.ts#L81-L93
+ */
+const videosAddResumableInitValidator = getCommonVideoEditAttributes().concat([
+  body('name')
+    .trim()
+    .custom(isVideoNameValid)
+    .withMessage('Should have a valid name'),
+  body('channelId')
+    .customSanitizer(toIntOrNull)
+    .custom(isIdValid).withMessage('Should have correct video channel id'),
+  header('x-upload-content-length')
+    .isNumeric()
+    .exists()
+    .withMessage('Should specify the file length'),
+  header('x-upload-content-type')
+    .isString()
+    .exists()
+    .withMessage('Should specify the file mimetype'),
+
+  async (req: express.Request, res: express.Response, next: express.NextFunction) => {
+    if (req.method !== 'POST') return next()
+    if (req.body.isPreviewForAudio) return next() // don't check mimetype for audio preview TODO: check size?
+
+    const videoFileMetadata = {
+      mimetype: req.headers['x-upload-content-type'],
+      size: +req.headers['x-upload-content-length'],
+      originalname: req.body.name
+    } as express.FileUploadMetadata
+    const user = res.locals.oauth.token.User
+    const file: File & { duration: number, path: string, filename: string } = req.body
+    file.path = getResumableUploadPath(file.id)
+
+    logger.debug('Checking videosAddResumableInitValidator parameters and headers', { parameters: req.body, headers: req.headers })
+
+    if (areValidationErrors(req, res)) return clearUploadFile(file.path)
+    if (areErrorsInScheduleUpdate(req, res)) return clearUploadFile(file.path)
+
+    if (!await doesVideoChannelOfAccountExist(req.body.channelId, user, res)) return clearUploadFile(file.path)
+
+    if (!isVideoFileMimeTypeValid({ videofile: [ videoFileMetadata ] })) {
+      res.status(HttpStatusCode.UNSUPPORTED_MEDIA_TYPE_415)
+         .json({
+           error: 'This file is not supported. Please, make sure it is of the following type: ' +
+                  CONSTRAINTS_FIELDS.VIDEOS.EXTNAME.join(', ')
+         })
+
+      return clearUploadFile(file.path)
+    }
+
+    if (!isVideoFileSizeValid(videoFileMetadata.size.toString())) {
+      res.status(HttpStatusCode.PAYLOAD_TOO_LARGE_413)
+         .json({ error: 'This file is too large.' })
+
+      return clearUploadFile(file.path)
+    }
+
+    if (await isAbleToUploadVideo(user.id, videoFileMetadata.size) === false) {
```

Comment:
What do you say about refactoring those validators to middlewares (instead of helpers)? Then it would look like something below (I think the function name should be renamed though). That makes it easier to get an overview about whats validated.

```
videosRouter.post('/upload',
  authenticate,
  reqVideoFileAdd,
  isAbleToUploadVideo,
  areErrorsInScheduleUpdate,
  doesVideoChannelOfAccountExist,
  asyncMiddleware(videosAddLegacyValidator),
  asyncRetryTransactionMiddleware(addVideoLegacy)
)

videosRouter.all('/upload-resumable',
  onlyAllowMethods([ 'POST', 'PUT', 'DELETE' ]), // uploadx also allows GET and PATCH
  authenticate,
  executeIfPOST(asyncMiddleware(videosAddResumableInitValidator)),
  uploadx.upload(uploadxOptions), // uploadx doesn't use call next() before the file upload completes
  isAbleToUploadVideo,
  areErrorsInScheduleUpdate,
  doesVideoChannelOfAccountExist,
  asyncMiddleware(videosAddResumableValidator),
  asyncMiddleware(addVideoResumable)
)
```

Reply:
Both methods seems fine but I prefer to use helpers for consistency (so we don't have to also update the update endpoint etc).